homeassistant/components/ihc/service_functions.py

```diff
@@ -58,9 +59,13 @@ def setup_service_functions(hass: HomeAssistant):
     """Set up the IHC service functions."""
 
     def _get_controller(call):
-        controller_id = call.data[ATTR_CONTROLLER_ID]
-        ihc_key = f"ihc{controller_id}"
-        return hass.data[ihc_key][IHC_CONTROLLER]
+        controller_index = call.data[ATTR_CONTROLLER_ID]
+        for controller_id in hass.data[DOMAIN]:
+            controller_conf = hass.data[DOMAIN][controller_id]
+            if controller_conf[IHC_CONTROLLER_INDEX] == controller_index:
+                return controller_conf[IHC_CONTROLLER]
+        # if not found the controller_index is ouf of range
+        _LOGGER.error("The controller index is of of range")
```

Comment:
Nowadays we normally raise an exception with a message as argument in service calls when there's a problem. The message will be shown in the frontend.

If the user service input is incorrect we raise a ValueError other a HomeAssistantError.